app/scripts/migrations/135.ts

```diff
@@ -0,0 +1,187 @@
+import { hasProperty, isObject } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 135;
+
+/**
+ * Removes tokens with `decimals === null` from `allTokens`, `allDetectedTokens`, `tokens`, and `detectedTokens`.
+ * Captures exceptions for invalid states using Sentry and logs tokens with `decimals === null`.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly
+ * what we persist to disk.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+/**
+ * Transforms the TokensController state to remove tokens with `decimals === null`.
+ *
+ * @param state - The persisted MetaMask state.
+ */
+function transformState(state: Record<string, unknown>): void {
+  if (!hasProperty(state, 'TokensController')) {
+    return;
+  }
+
+  const tokensControllerState = state.TokensController;
+
+  if (!isObject(tokensControllerState)) {
+    global.sentry?.captureException(
+      new Error(
+        `Migration ${version}: Invalid TokensController state of type '${typeof tokensControllerState}'`,
+      ),
+    );
+    return;
+  }
+
+  // Validate and transform `allTokens`
+  if (hasProperty(tokensControllerState, 'allTokens')) {
+    if (!isObject(tokensControllerState.allTokens)) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: Invalid allTokens state of type '${typeof tokensControllerState.allTokens}'`,
+        ),
+      );
+    } else {
+      tokensControllerState.allTokens = transformTokenCollection(
+        tokensControllerState.allTokens,
+        'allTokens',
+      );
+    }
+  }
+
+  // Validate and transform `allDetectedTokens`
+  if (hasProperty(tokensControllerState, 'allDetectedTokens')) {
+    if (!isObject(tokensControllerState.allDetectedTokens)) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: Invalid allDetectedTokens state of type '${typeof tokensControllerState.allDetectedTokens}'`,
+        ),
+      );
+    } else {
+      tokensControllerState.allDetectedTokens = transformTokenCollection(
+        tokensControllerState.allDetectedTokens,
+        'allDetectedTokens',
+      );
+    }
+  }
+
+  // Transform `tokens` array
+  if (
+    hasProperty(tokensControllerState, 'tokens') &&
+    Array.isArray(tokensControllerState.tokens)
+  ) {
+    tokensControllerState.tokens = tokensControllerState.tokens.filter(
+      (token) => {
+        if (
+          isObject(token) &&
+          hasProperty(token, 'decimals') &&
+          token.decimals === null &&
```

Comment:
Do we know if the bug can be trigger by decimals being `undefined`?

Reply:
yes we did the test with @danjm  for this case , it's not triggering the error